massdatasets.py

```diff
@@ -41,22 +46,47 @@ def add_song(self, artist, title, style, audio_filepath, **kwargs):
              **kwargs}
         )
 
-    def to_pandas_df(self):
+    def to_pandas_df(self, include_features=True):
         '''
         Compiles the yaml document to a pandas DataFrame.
         audio_filepath are complete (prefixed by base_path).
         '''
 
-        frame = pd.DataFrame(columns=self.songs[0].keys())
+        long_format = True
+        features = []
+        frames = []
+        songs = self.songs.copy()
+
+        for song in songs:
+            if 'feature' in song and include_features:
+                features.append(pd.DataFrame.from_dict(song.pop('feature')))
+            frames.append(pd.DataFrame.from_dict(song))
 
-        for song in self.songs:
-            sub_frame = pd.DataFrame.from_dict(song)
-            sub_frame['audio'] = sub_frame.index
-            frame = frame.append(sub_frame, ignore_index=True)
+        frame = pd.concat(frames, copy=False).reset_index()
+        frame.rename(columns={'index': 'source'}, inplace=True)
 
         frame['audio_filepath'] = ['/'.join((self.base_path, _))
                                    for _ in frame['audio_filepath']]
         frame['dataset'] = self.dataset
+
+        # This is clunky way to add features in long format
+        if features and include_features:
+            frame2 = pd.concat(features, copy=False)
+            cols1, cols2 = list(frame.columns), list(frame2.columns)
```

Comment:
`cols1` and `cols2` is only used in the if-statement below, so you could move this line inside that statement.

Reply:
Hopefully improved readability somewhat now